File: frontend-design/scripts/generate_slide_brief.py
#!/usr/bin/env python3
"""Generate a slide/presentation or section-layout brief from slide data.

Usage:
  python scripts/generate_slide_brief.py "investor pitch traction"
  python scripts/generate_slide_brief.py "problem slide" --position 2 --total 10
"""

from __future__ import annotations

import argparse
import csv
import re
from pathlib import Path
from math import log

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data"


def toks(text: str) -> list[str]:
    return re.findall(r"[a-zA-Z0-9_+-]+", text.lower())


def score(query: list[str], row: dict[str, str]) -> float:
    words = toks(" ".join(row.values()))
    if not words:
        return 0.0
    norm = 1 + log(len(words) + 1)
    return sum((words.count(q) + sum(0.2 for w in set(words) if q in w and q != w)) / norm for q in query)


def load(name: str) -> list[dict[str, str]]:
    with (DATA / name).open(newline="", encoding="utf-8") as handle:
        return list(csv.DictReader(handle))


def top(name: str, query: list[str]) -> dict[str, str] | None:
    ranked = sorted(((score(query, row), row) for row in load(name)), key=lambda item: item[0], reverse=True)
    return ranked[0][1] if ranked and ranked[0][0] > 0 else None


def val(row: dict[str, str] | None, *keys: str) -> str:
    if not row:
        return "—"
    for key in keys:
        if row.get(key):
            return row[key]
    return "—"


def main() -> int:
    parser = argparse.ArgumentParser(description="Generate slide/section brief")
    parser.add_argument("query", nargs="+")
    parser.add_argument("--position", type=int)
    parser.add_argument("--total", type=int)
    args = parser.parse_args()
    query_text = " ".join(args.query)
    query = toks(query_text)
    strategy = top("slide-strategies.csv", query)
    layout = top("slide-layouts.csv", query)
    logic = top("slide-layout-logic.csv", query)
    copy = top("slide-copy.csv", query)
    color = top("slide-color-logic.csv", query)
    typography = top("slide-typography.csv", query)
    chart = top("slide-charts.csv", query)
    background = top("slide-backgrounds.csv", query)
    print(f"# Slide / Section Brief: {query_text}\n")
    if args.position and args.total:
        break_pattern = args.position in {max(1, args.total // 3), max(1, (args.total * 2) // 3)}
        print(f"- Position: {args.position}/{args.total}")
        print(f"- Pattern break candidate: {'yes' if break_pattern else 'no'}")
    print(f"- Strategy: {val(strategy, 'strategy_name')} — {val(strategy, 'goal')}")
    print(f"- Layout: {val(layout, 'layout_name')} — {val(layout, 'use_case')}")
    print(f"- Logic: {val(logic, 'layout_pattern')} / {val(logic, 'direction')} / visual weight {val(logic, 'visual_weight')}")
    print(f"- Copy formula: {val(copy, 'formula_name')} — {val(copy, 'example_template')}")
    print(f"- Color: bg {val(color, 'background')}, text {val(color, 'text_color')}, accent {val(color, 'accent_usage')}")
    print(f"- Typography: primary {val(typography, 'primary_size')}, secondary {val(typography, 'secondary_size')}, line-height {val(typography, 'line_height')}")
    print(f"- Chart if needed: {val(chart, 'chart_type')} — {val(chart, 'when_to_use')}")
    print(f"- Background: {val(background, 'image_category')} with {val(background, 'overlay_style')} overlay, text {val(background, 'text_placement')}")
    print("\n## Rules\n")
    print("- One idea per slide/viewport.")
    print("- Prefer direct labels and large readable type.")
    print("- Pattern breaks are useful around one-third and two-thirds of a long sequence.")
    print("- Charts need labels, not just visual shape.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: frontend-design/scripts/generate_slide_brief.mjs
#!/usr/bin/env node
/** Generate a slide, presentation, or section-layout brief. */

import { pathToFileURL } from "node:url";
import { loadCsv, pick, rowText, score, tokenize } from "./_catalog.mjs";

function parseArgs(args) {
  let position;
  let total;
  const query = [];
  for (let index = 0; index < args.length; index += 1) {
    if (args[index] === "--position") {
      position = Number.parseInt(args[index + 1], 10);
      index += 1;
    } else if (args[index] === "--total") {
      total = Number.parseInt(args[index + 1], 10);
      index += 1;
    } else query.push(args[index]);
  }
  if (query.length === 0) throw new Error("Usage: generate_slide_brief.mjs <query> [--position <n>] [--total <n>]");
  return { query: query.join(" "), position, total };
}

function top(name, query) {
  const ranked = loadCsv(name).map((row) => [score(query, rowText(row)), row]).sort((left, right) => right[0] - left[0]);
  return ranked.length > 0 && ranked[0][0] > 0 ? ranked[0][1] : undefined;
}

function value(row, ...keys) {
  return pick(row, ...keys) || "—";
}

export function main(args = process.argv.slice(2)) {
  let options;
  try {
    options = parseArgs(args);
  } catch (error) {
    console.error(error.message);
    return 2;
  }
  const query = tokenize(options.query);
  const strategy = top("slide-strategies.csv", query);
  const layout = top("slide-layouts.csv", query);
  const logic = top("slide-layout-logic.csv", query);
  const copy = top("slide-copy.csv", query);
  const color = top("slide-color-logic.csv", query);
  const typography = top("slide-typography.csv", query);
  const chart = top("slide-charts.csv", query);
  const background = top("slide-backgrounds.csv", query);
  const lines = [`# Slide / Section Brief: ${options.query}`, ""];
  if (options.position && options.total) {
    const breakpoints = new Set([Math.max(1, Math.floor(options.total / 3)), Math.max(1, Math.floor(options.total * 2 / 3))]);
    lines.push(`- Position: ${options.position}/${options.total}`, `- Pattern break candidate: ${breakpoints.has(options.position) ? "yes" : "no"}`);
  }
  lines.push(
    `- Strategy: ${value(strategy, "strategy_name")} — ${value(strategy, "goal")}`,
    `- Layout: ${value(layout, "layout_name")} — ${value(layout, "use_case")}`,
    `- Logic: ${value(logic, "layout_pattern")} / ${value(logic, "direction")} / visual weight ${value(logic, "visual_weight")}`,
    `- Copy formula: ${value(copy, "formula_name")} — ${value(copy, "example_template")}`,
    `- Color: bg ${value(color, "background")}, text ${value(color, "text_color")}, accent ${value(color, "accent_usage")}`,
    `- Typography: primary ${value(typography, "primary_size")}, secondary ${value(typography, "secondary_size")}, line-height ${value(typography, "line_height")}`,
    `- Chart if needed: ${value(chart, "chart_type")} — ${value(chart, "when_to_use")}`,
    `- Background: ${value(background, "image_category")} with ${value(background, "overlay_style")} overlay, text ${value(background, "text_placement")}`,
    "", "## Rules", "",
    "- One idea per slide/viewport.",
    "- Prefer direct labels and large readable type.",
    "- Pattern breaks are useful around one-third and two-thirds of a long sequence.",
    "- Charts need labels, not just visual shape.",
  );
  process.stdout.write(`${lines.join("\n")}\n`);
  return 0;
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) process.exitCode = main();
